dimensiondata/src/test/java/org/jclouds/dimensiondata/cloudcontrol/compute/function/ServerToServerWithExternalIpTest.java

```diff
@@ -0,0 +1,118 @@
+      networkApi = EasyMock.createNiceMock(NetworkApi.class);
+      natRule = EasyMock.createNiceMock(NatRule.class);
+
+      nic = EasyMock.createNiceMock(NIC.class);
+      cpu = EasyMock.createNiceMock(CPU.class);
+      os = OperatingSystem.builder().family(OsFamily.WINDOWS.name()).id("Win10").displayName("Windows 10").build();
+
+      EasyMockSupport.injectMocks(dimensionDataCloudControlApi);
+      EasyMockSupport.injectMocks(networkApi);
+   }
+
+   @Test
+   public void testServerToServerWithExternalIpApplyNotNull() {
+      String internalIp = "192.168.1.1";
+      String externalIp = "10.12.122.1";
+      String networkDomainId = "NetworkDomain1";
+
+      server = Server.builder().id("serverId").name("serverName").datacenterId("NA1")
+            .networkInfo(NetworkInfo.create(networkDomainId, nic, new ArrayList<NIC>())).cpu(cpu).deployed(true)
+            .state(State.NORMAL).sourceImageId("imageId").started(false).createTime(new Date()).memoryGb(1024)
+            .guest(Guest.builder().osCustomization(false).operatingSystem(os).build()).build();
+
+      PagedIterable<NatRule> natRules = PagedIterables.onlyPage(IterableWithMarkers.from(Lists.newArrayList(natRule)));
+
+      expect(dimensionDataCloudControlApi.getNetworkApi()).andReturn(networkApi);
+      expect(networkApi.listNatRules(networkDomainId)).andReturn(natRules);
+      expect(nic.privateIpv4()).andReturn(internalIp);
+      expect(natRule.externalIp()).andReturn(externalIp);
+      expect(natRule.internalIp()).andReturn(internalIp);
+
+      EasyMock.replay(dimensionDataCloudControlApi, networkApi, natRule, nic);
+
+      ServerWithExternalIp result = new ServerToServerWithExternalIp(dimensionDataCloudControlApi).apply(server);
+      assertNotNull(result);
+      assertEquals(result.server(), server);
+      assertEquals(result.externalIp(), externalIp);
+   }
+
+   @Test(dependsOnMethods = "testServerToServerWithExternalIpApplyNotNull")
+   public void testServerToServerWithExternalIpApplyNull() {
+      server = null;
+      ServerWithExternalIp result = new ServerToServerWithExternalIp(dimensionDataCloudControlApi).apply(server);
+      assertNull(result);
+   }
+
+   @Test(dependsOnMethods = "testServerToServerWithExternalIpApplyNotNull")
+   public void testServerToServerWithExternalIpApplyNetworkInfoNull() {
+      server = Server.builder().id("serverId").name("serverName").datacenterId("NA1").networkInfo(null).cpu(cpu)
+            .deployed(true).state(State.NORMAL).sourceImageId("imageId").started(false).createTime(new Date())
+            .memoryGb(1024).guest(Guest.builder().osCustomization(false).operatingSystem(os).build()).build();
+
+      ServerWithExternalIp result = new ServerToServerWithExternalIp(dimensionDataCloudControlApi).apply(server);
+      assertNotNull(result);
+      assertEquals(result.server(), server);
+      assertNull(result.externalIp());
+   }
```

Comment:
Add a test case to verify what happens when there is network info but no matching nat rule is found.